aten/src/ATen/Utils.h

```diff
@@ -127,4 +128,25 @@ inline int64_t prod_intlist(ArrayRef<int64_t> list) {
   return std::accumulate(list.begin(), list.end(), 1ll, std::multiplies<int64_t>());
 }
 
+/**
+ * Utility function used in tensor implementations, which
+ * supplies the default generator to tensors, if an input generator
+ * is not supplied. The input Generator* is also dynamic casted to
+ * the backend generator type (CPU/CUDAGenerator etc.)
+ * 
+ * See Note [Thread-safety and Generators]
+ */
+template <typename T>
+static inline T * check_generator_with_default(Generator * expr, Generator * defaultValue) {
+  if (!expr) {
+    expr = defaultValue;
+  }
+  // See Note [Thread-safety and Generators]
+  std::lock_guard<std::mutex> lock(expr->mutex_);
```

Comment:
This lock looks totally unnecessary (under the same reasoning: concurrent read-only operations to a generator should be intrinsically thread safe.)